### ISSUE TITLE
Custom UI node question

### ISSUE THREAD

**Issue by ksobon** [DynamoDS/Dynamo#6997]
Can someone give me an example of how I could bind the input (string) from the custom node to webbrowser? I am trying this on this Sample file: 
https://github.com/DynamoDS/DynamoSamples/blob/master/src/SampleLibraryUI/Examples/CustomNodeModel.cs

CC: @mjkkirschner

**Comment by ke-yu** [comment]
@ksobon it is a little bit tricky to get the real value of input.

Basic idea is to schedule an async task to Dynamo's `Scheduler` and to query the real data from the engine based on the variable name of input, in this task:
```
var runtimeMirror = engine.GetMirror(variableName);
if (runtimeMirror != null)
{
    var v = runtimeMirror.GetData().Data as string;
}
``` 

To schedule an async task, you may refer to [`ColorRange.UpdateColorRange()` implementation](https://github.com/DynamoDS/Dynamo/blob/master/src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/ColorRange.cs).

**Comment by mjkkirschner** [comment]
@ksobon exactly what @ke-yu said, you have to rely on the get engine and make sure to do null checking as you cannot be sure what the type will be or if any node is connected at all.

**Comment by ksobon** [comment]
OK, here's basically the SampleUI example that you guys posted a while ago with some small changes. @mjkkirschner you might be familiar with the idea to bind a WebBrowser to a NodeModel. I think I am getting closer on that: 

```C#
using Autodesk.DesignScript.Runtime;
using Dynamo.Controls;
using Dynamo.Graph.Nodes;
using Dynamo.UI.Commands;
using Dynamo.Wpf;
using System.ComponentModel;
using System.Windows;
using System.Windows.Media;

namespace SampleLibraryUI
{
    /// <summary>
    /// 
    /// </summary>
    [NodeName("Custom Node Model")]
    [NodeCategory("SampleLibraryUI.Examples")]
    [NodeDescription("CustomNodeModelDescription")]
    [IsDesignScriptCompatible]
    public class CustomNodeModel : NodeModel, INotifyPropertyChanged
    {
        /// <summary>
        /// 
        /// </summary>
        public static DynamoView dv;

        private string _myHtml;

        /// <summary>
        /// 
        /// </summary>
        public string MyHtml
        {
            get { return _myHtml; }
            set
            {
                if (_myHtml != value)
                {
                    _myHtml = value;
                    RaisePropertyChanged("MyHtml");
                    OnNodeModified();
                }
            }
        }

        private string message;

        /// <summary>
        /// 
        /// </summary>
        public string Message
        {
            get { return message; }
            set
            {
                message = value;
                RaisePropertyChanged("NodeMessage");
            }
        }

        /// <summary>
        /// 
        /// </summary>
        [IsVisibleInDynamoLibrary(false)]
        public DelegateCommand MessageCommand { get; set; }

        /// <summary>
        /// 
        /// </summary>
        public CustomNodeModel()
        {
            InPortData.Add(new PortData("htmlString", "Html String to render."));
            RegisterAllPorts();
            ArgumentLacing = LacingStrategy.Disabled;

            MessageCommand = new DelegateCommand(ShowMessage, CanShowMessage);
            Message = "Launch Window";
        }

        private static bool CanShowMessage(object obj)
        {
            return true;
        }

        private void ShowMessage(object obj)
        {
            MainWindow win = new MainWindow();
            win.Show();
        }
    }

    /// <summary>
    /// 
    /// </summary>
    public class CustomNodeModelNodeViewCustomization : INodeViewCustomization<CustomNodeModel>
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="model"></param>
        /// <param name="nodeView"></param>
        public void CustomizeView(CustomNodeModel model, NodeView nodeView)
        {
            var helloDynamoControl = new HelloDynamoControl();
            NodeView nv = nodeView;
            CustomNodeModel.dv = FindUpVisualTree<DynamoView>(nv);
            nodeView.inputGrid.Children.Add(helloDynamoControl);
            helloDynamoControl.DataContext = model;
        }

        // Thanks to Hans Hubers for this bit.
        private static T FindUpVisualTree<T>(DependencyObject initial) where T : DependencyObject
        {
            DependencyObject current = initial;
            while (current != null && current.GetType() != typeof(T))
            {
                current = VisualTreeHelper.GetParent(current);
            }
            return current as T;
        }

        /// <summary>
        /// 
        /// </summary>
        public void Dispose() { }
    }

}
```
So there is the <code> public string MyHtml </code> that is bound to the <code>WebBrowser</code> via attached property (example: https://thomasfreudenberg.com/archive/2010/08/01/binding-webbrowser-content-in-wpf/) I just can't figure out how to update that property when new string is fed through the input. Would I need to implement something similar to <code> ColorRange.UpdateColorRange() </code> ? You guys care for an example? 

Thanks! 

Just for context: MainWindow code behind: 

```C#
using System.Windows;

namespace SampleLibraryUI
{
    public partial class MainWindow : Window
    {
        /// <summary>
        /// 
        /// </summary>
        public MainWindow()
        {
            InitializeComponent();

            var model = new CustomNodeModel()
            {
                MyHtml = @"
<html>
  <body>
    Some sample HTML code.
  </body>
</html>"
            };

            DataContext = model;
            Owner = CustomNodeModel.dv;
        }
    }
}
```
MainWindow xmal:
```
<Window x:Class="SampleLibraryUI.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" 
        xmlns:src="clr-namespace:SampleLibraryUI" 
        Title="SampleLibraryUI" Height="350" Width="525">
    <Grid>
        <WebBrowser
            src:WebBrowserHelper.Body="{Binding MyHtml}"
            />
    </Grid>
</Window>
```
WebBrowserHelper method that adds an attached property to WebBrowser: 

```
using System.Windows;
using System.Windows.Controls;

namespace SampleLibraryUI
{
    class WebBrowserHelper
    {
        public static readonly DependencyProperty BodyProperty =
            DependencyProperty.RegisterAttached("Body", typeof(string), typeof(WebBrowserHelper), new PropertyMetadata(OnBodyChanged));

        public static string GetBody(DependencyObject dependencyObject)
        {
            return (string)dependencyObject.GetValue(BodyProperty);
        }

        public static void SetBody(DependencyObject dependencyObject, string body)
        {
            dependencyObject.SetValue(BodyProperty, body);
        }

        private static void OnBodyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            var webBrowser = (WebBrowser)d;
            webBrowser.NavigateToString((string)e.NewValue);
        }
    }
}
```

**Comment by ksobon** [comment]
Here's a downloadable project: https://drive.google.com/folderview?id=0B3QQaPkepGSkRzE0WDdIWjM0ak0&usp=sharing

**Comment by ksobon** [comment]
I tried creating an asynch task like @ke-yu was suggesting. I can see that when <code>UpdateHtmlString()</code> is called the proper input string is being retrieved. I tried assigning it to the <code>CustomNodeModel.MyHtml</code> property right from that method, but I don't see any changes happening to the WebBrowser. I am not sure if its the binding issue or I am scheduling the task wrong. Any help here will be appreciated. 

Code: 

```C#
using Autodesk.DesignScript.Runtime;
using Dynamo.Controls;
using Dynamo.Graph.Nodes;
using Dynamo.UI.Commands;
using Dynamo.Wpf;
using System;
using System.ComponentModel;
using System.Windows;
using System.Windows.Media;
using Dynamo.Models;
using Dynamo.ViewModels;
using Dynamo.Scheduler;
using Dynamo.Engine;
using System.Linq;

namespace SampleLibraryUI
{
    /// <summary>
    /// 
    /// </summary>
    [NodeName("Custom Node Model")]
    [NodeCategory("SampleLibraryUI.Examples")]
    [NodeDescription("CustomNodeModelDescription")]
    [IsDesignScriptCompatible]
    public class CustomNodeModel : NodeModel, INotifyPropertyChanged
    {
        public event Action RequestChangeHtmlString;
        protected virtual void OnRequestChangeHtmlString()
        {
            if (RequestChangeHtmlString != null)
                RequestChangeHtmlString();
        }

        public static DynamoView dv;

        private string _myHtml;
        public string MyHtml
        {
            get { return _myHtml; }
            set
            {
                if (_myHtml != value)
                {
                    _myHtml = value;
                    RaisePropertyChanged("MyHtml");
                }
            }
        }

        private string message;
        public string Message
        {
            get { return message; }
            set
            {
                message = value;
                RaisePropertyChanged("NodeMessage");
            }
        }

        [IsVisibleInDynamoLibrary(false)]
        public DelegateCommand MessageCommand { get; set; }

        /// <summary>
        /// 
        /// </summary>
        public CustomNodeModel()
        {
            InPortData.Add(new PortData("htmlString", "Html String to render."));
            RegisterAllPorts();
            ArgumentLacing = LacingStrategy.Disabled;

            this.PropertyChanged += HtmlString_PropertyChanged;
            foreach (var port in InPorts)
            {
                port.Connectors.CollectionChanged += Connectors_CollectionChanged;
            }

            MessageCommand = new DelegateCommand(ShowMessage, CanShowMessage);
            Message = "Launch Window";
        }

        void Connectors_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            OnRequestChangeHtmlString();
        }

        void HtmlString_PropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (e.PropertyName != "CachedValue")
                return;

            if (InPorts.Any(x => x.Connectors.Count == 0))
                return;

            OnRequestChangeHtmlString();
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="engine"></param>
        /// <returns></returns>
        public string GetInputString(EngineController engine)
        {
            string htmlString;

            // If there are colors supplied
            if (HasConnectedInput(0))
            {
                var htmlNode = InPorts[0].Connectors[0].Start.Owner;
                var htmlIndex = InPorts[0].Connectors[0].Start.Index;
                var nodeName = htmlNode.GetAstIdentifierForOutputIndex(htmlIndex).Name;
                var htmlMirror = engine.GetMirror(nodeName);
                htmlString = htmlMirror.GetData().Data as string;
            }
            else
            {
                htmlString = string.Empty;
            }

            this.MyHtml = htmlString;
            return htmlString;
        }

        private static bool CanShowMessage(object obj)
        {
            return true;
        }

        private void ShowMessage(object obj)
        {
            MainWindow win = new MainWindow();
            win.Show();
        }
    }

    /// <summary>
    /// 
    /// </summary>
    public class CustomNodeModelNodeViewCustomization : INodeViewCustomization<CustomNodeModel>
    {
        private DynamoModel dynamoModel;
        private DynamoViewModel dynamoViewModel;
        private CustomNodeModel mandrillNode;

        /// <summary>
        /// 
        /// </summary>
        /// <param name="model"></param>
        /// <param name="nodeView"></param>
        public void CustomizeView(CustomNodeModel model, NodeView nodeView)
        {
            this.dynamoModel = nodeView.ViewModel.DynamoViewModel.Model;
            this.dynamoViewModel = nodeView.ViewModel.DynamoViewModel;
            mandrillNode = model;

            var helloDynamoControl = new HelloDynamoControl();
            NodeView nv = nodeView;
            CustomNodeModel.dv = FindUpVisualTree<DynamoView>(nv);
            nodeView.inputGrid.Children.Add(helloDynamoControl);
            helloDynamoControl.DataContext = model;

            mandrillNode.RequestChangeHtmlString += UpdateHtmlString;
        }

        private void UpdateHtmlString()
        {
            var s = dynamoViewModel.Model.Scheduler;
            var t = new DelegateBasedAsyncTask(s, () =>
            {
                mandrillNode.MyHtml = mandrillNode.GetInputString(dynamoModel.EngineController);
            });

            s.ScheduleForExecution(t);
        }

        // Thanks to Hans Hubers for this bit.
        private static T FindUpVisualTree<T>(DependencyObject initial) where T : DependencyObject
        {
            DependencyObject current = initial;
            while (current != null && current.GetType() != typeof(T))
            {
                current = VisualTreeHelper.GetParent(current);
            }
            return current as T;
        }

        public void Dispose() { }
    }

}
```

**Comment by mjkkirschner** [comment]
so, I'm a bit confused, it looks like you are creating a `customNodeModel` instance when the main window is shown, and the window is shown when a first `customNodeModel` is created... I think maybe the `DataContext` is not set correctly, so the binding is not working.

Seems you could pass the node model to the window when you construct as one way to set the data context.

**Comment by ksobon** [comment]
Yes, you are right. Passing the model to the control as its created and setting its context fixes the binding. Now it works nicely. 

```C#
public void CustomizeView(CustomNodeModel model, NodeView nodeView)
        {
            this.dynamoModel = nodeView.ViewModel.DynamoViewModel.Model;
            this.dynamoViewModel = nodeView.ViewModel.DynamoViewModel;
            mandrillNode = model;

            var helloDynamoControl = new HelloDynamoControl();
            NodeView nv = nodeView;
            CustomNodeModel.dv = FindUpVisualTree<DynamoView>(nv);
            nodeView.inputGrid.Children.Add(helloDynamoControl);
            helloDynamoControl.DataContext = model;

            var mandrillWindow = new MainWindow();
            mandrillWindow.DataContext = model;
            mandrillWindow.Show();

            mandrillNode.RequestChangeHtmlString += UpdateHtmlString;
        }
```

**Comment by mjkkirschner** [comment]
@ksobon I'm glad we could finally solve this almost one year later 😉

**Comment by ksobon** [comment]
oh no, I had it hacked together for a while but it wasn't properly built (it was hacked together with some events and new thread) so it kept crashing Dynamo when I tried running from Revit. It worked great on top of sandbox though so I was able to build up some more functionality for it in the meantime I am trying to get it fixed up properly before publishing. :-) 

Now, I just need to figure out how to check if window is already open and not open it when button is being pressed on the control. I might be posting more questions soon...thanks!
